kong/runloop/handler.lua

```diff
@@ -721,25 +718,86 @@ return {
 
       local trusted_ip = kong.ip.is_trusted(realip_remote_addr)
       if trusted_ip then
-        forwarded_proto = var.http_x_forwarded_proto or var.scheme
-        forwarded_host  = var.http_x_forwarded_host  or var.host
-        forwarded_port  = var.http_x_forwarded_port  or var.server_port
+        forwarded_proto = var.http_x_forwarded_proto          or scheme
+        forwarded_host  = var.http_x_forwarded_host           or host
+        forwarded_port  = tonumber(var.http_x_forwarded_port) or port
 
       else
-        forwarded_proto = var.scheme
-        forwarded_host  = var.host
-        forwarded_port  = var.server_port
+        forwarded_proto = scheme
+        forwarded_host  = host
+        forwarded_port  = port
       end
 
-      local protocols = route.protocols
-      if (protocols and
-          protocols.https and not protocols.http and forwarded_proto ~= "https")
+      local protocols = route and route.protocols
+      if (protocols and protocols.https and not protocols.http and
+          forwarded_proto ~= "https")
       then
         ngx.header["connection"] = "Upgrade"
         ngx.header["upgrade"]    = "TLS/1.2, HTTP/1.1"
         return kong.response.exit(426, { message = "Please use HTTPS protocol" })
       end
 
+      -- Service-less HTTP Route
+      if not service then
+        local service_scheme
+        local service_host
+        local service_port
+
+        local http_host = var.http_host
+        if http_host then
+          http_host = lower(http_host)
+
+          -- Micro-optimization to make sure we use JITable functions in this
+          -- potentially hot path, e.g. this simple code doesn't warrant us
+          -- using something like socket.url.parse:
+          -- https://github.com/diegonehab/luasocket/blob/043e99771352aff47680b99f09b66a32f0cc3ef5/src/url.lua#L143-L192
+          local s = find(http_host, ":", 2, true)
+          if s then
+            local p = tonumber(sub(http_host, s + 1))
```

Comment:
Pass base to `tonumber`?

Reply:
Passing it now. Thank you @james-callahan for the review!